### ISSUE TITLE
Overly conservative nullable analysis for finally blocks

### ISSUE THREAD

**Issue by sbomer** [dotnet/roslyn#58057]
**Steps to Reproduce**:
https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzggPgYhgVwDZ4TB4CmABCTEaQLABQAwgPxkAeA3PQgE4CeZAb3oBfMgDNYEAvyF0y8tmQC8ZXAU5yFrZapIB3MgwAUASg3D6ANwjcyKHawB0AFQD2AZR6wA5qY30AAQAmQ0FhIA===

```csharp
C? x;
try {
} finally {
    x = null;
    x = new C();
}
var s = x.ToString(); // nullable warning
```

**Expected Behavior**:
This shouldn't warn because the variable is always assigned a non-null value at the exit of the finally block.

**Actual Behavior**:
It produces a warning.

**Comment by RikkiGibson** [comment]
@jcouv do you remember why this is the case?

**Comment by agocke** [comment]
I can't remember why we might need this for region analysis, but my general suggestions is to delete the line and see which tests fail 🙃

**Comment by jcouv** [comment]
For the `try` block, we assume that every statement could throw. So an assignment to a maybe-null value followed by an assignment to a not-null value will still produce a maybe-null as the starting point for the `catch` and `finally` blocks ([example](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzggPgYhgVwDZ4TB4CmABCTEaQLABQ9AwgPxkAeZAvGTCQO5lGACgCUAbnr0EAJwCeZegG96ZVey498eCXTXruvAcPH0AvvQBmsCAXlKVagPoa2AOgAqAewDKM2AHNRHXMGOgABACZBMkVTIA==)).
That logic is handled in [`NullableWalker.SetStateAndTrackForFinally`](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs#L2281) and [`AbstractFlowPass.VisitTryBlockWithAnyTransferFunction`](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs#L1724).

The surprising part based on [semantics](https://github.com/dotnet/csharplang/blob/main/spec/statements.md#the-try-statement) of `try` statement is that [`AbstractFlowPass.VisitFinallyBlockWithAnyTransferFunction`](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs#L1798) which analyzes the `finally` block does the same thing (it sets a `NonMonotonicState`, then visit the statements, and so `SetStateAndTrackForFinally` computes a state assuming jumps to the end of the block after each assignment).
The effect of the `finally`, computed by `VisitFinallyBlockWithAnyTransferFunction`, has `x` as maybe-null...
